Moves TestNativeHistogram_MulDivOperator to promql testing framework

NeerajGartia21 opened the pull request:
<!--
    Please give your PR a title in the form "area: short description".  For example "tsdb: reduce disk usage by 95%"

    If your PR is to fix an issue, put "Fixes #issue-number" in the description.

    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->
related to: #13563 
Also this PR adds support of Inf and NaN as numbers in histogram buckets
@beorn7 @fatsheep9146 please review !

beorn7 commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -718,6 +718,48 @@ eval instant at 10m histogram_fraction(-Inf, +Inf, histogram_fraction_4)
 eval instant at 10m histogram_sum(scalar(histogram_fraction(-Inf, +Inf, sum(histogram_fraction_4))) * histogram_fraction_4)
     {} 100
 
+# Apply multiplication and division operator to histogram
```
Nit: End sentence with period.

NeerajGartia21 replied:
done !

beorn7 commented on promql/parser/lex.go:
```diff
@@ -617,6 +617,16 @@ func lexBuckets(l *Lexer) stateFn {
 		l.bracketOpen = false
 		l.emit(RIGHT_BRACKET)
 		return lexHistogram
+	case isAlpha(r):
```
Could you add a test case for this to lex_test.go?

NeerajGartia21 replied:
added!

beorn7 reviewed (changes requested):
Thank you very much. Looks good in general.

I'll check out the tests in detail once #14585 is merged and we can actually run the tests.

In the meantime:
- Let's add a test case for the changed lexer (there are many cases already in lex_test.go, should be very easy to add a buckets with NaN and Inf).
- Please also remove the converted test `TestNativeHistogram_MulDivOperator`.

NeerajGartia21 commented:
> I'll check out the tests in detail once https://github.com/prometheus/prometheus/pull/14585 is merged and we can actually run the tests.

Cool. I've added the tests for changes in lexer and removed the added tests from `engine.go` file.

beorn7 commented:
#14585 is merged. Please rebase this so that we can see if the tests pass.

NeerajGartia21 commented:
Rebased! And All tests are passed!

beorn7 commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -718,6 +718,48 @@ eval instant at 10m histogram_fraction(-Inf, +Inf, histogram_fraction_4)
 
+# Apply multiplication and division operator to histogram.
+load 10m
+    histogram_mul_div {{schema:0 count:21 sum:33 z_bucket:3 z_bucket_w:0.001 buckets:[3 3 3] n_buckets:[3 3 3]}}x1
+	float_series_3 3+0x1
+	float_series_0 0+0x1
```
Some lines use blanks and some lines use tabs.

At least within the same test, we should be consistent. I recommend to use just blanks everywhere (your editor probably has an "untabify" function – it'l literally called `untabify` in emacs, for example).

NeerajGartia21 replied:
done!

beorn7 commented on promql/promqltest/testdata/native_histograms.test:
```diff
@@ -718,6 +718,48 @@ eval instant at 10m histogram_fraction(-Inf, +Inf, histogram_fraction_4)
+eval instant at 10m float_series_0*histogram_mul_div
+    {} {{schema:0 count:0 sum:0 z_bucket:0 z_bucket_w:0.001 buckets:[0 0 0] n_buckets:[0 0 0]}}
+
+eval instant at 10m histogram_mul_div/0
+    {} {{schema:0 count:Inf sum:Inf z_bucket:Inf z_bucket_w:0.001 buckets:[Inf Inf Inf] n_buckets:[Inf Inf Inf]}}
+
+eval instant at 10m histogram_mul_div/float_series_0
+    {} {{schema:0 count:Inf sum:Inf z_bucket:Inf z_bucket_w:0.001 buckets:[Inf Inf Inf] n_buckets:[Inf Inf Inf]}}
```
Let's add a TODO here that we plan to change this behavior (linking to #13934).

NeerajGartia21 replied:
Done!

beorn7 reviewed (changes requested):
Looks great. Just a tab vs. spaces nit and a request for a comment.

And one more thing: Maybe add a new test (now that it is so easy) evaluating `histogram_mul_div*0/0` (to see `NaN`s popping up).

beorn7 commented:
Also, we have a merge conflict now…

NeerajGartia21 commented:
Resolved all the merge conflicts and applied all the suggested changes!
